health-checker/src/main/java/io/stargate/health/StorageHealthChecker.java

```diff
@@ -0,0 +1,158 @@
+
+  private static final String PK_COLUMN_NAME = "pk";
+  private static final String VALUE_COLUMN_NAME = "value";
+  private static final String VALUE_COLUMN_VALUE = "dummy_value";
+  private static final Table EXPECTED_TABLE =
+      ImmutableTable.builder()
+          .keyspace(KEYSPACE_NAME)
+          .name(TABLE_NAME)
+          .addColumns(
+              ImmutableColumn.create(PK_COLUMN_NAME, Column.Kind.PartitionKey, Column.Type.Uuid),
+              ImmutableColumn.create(VALUE_COLUMN_NAME, Column.Kind.Regular, Column.Type.Varchar))
+          .build();
+  private final DataStoreFactory dataStoreFactory;
+
+  public StorageHealthChecker(DataStoreFactory dataStoreFactory)
+      throws ExecutionException, InterruptedException {
+
+    this.dataStoreFactory = dataStoreFactory;
+    if (SHOULD_CREATE_KS_AND_TABLE) {
+      ensureTableExists(dataStoreFactory.createInternal());
+    }
+  }
+
+  @Override
+  protected Result check() throws Exception {
+    if (!STORAGE_CHECK_ENABLED) {
+      return Result.healthy("Storage check disabled");
+    }
+    try {
+      DataStore dataStore = dataStoreFactory.createInternal();
+
+      Instant writeTimestamp = Instant.now();
+      // insert record
+      dataStore
+          .execute(
+              dataStore
+                  .queryBuilder()
+                  .insertInto(KEYSPACE_NAME, TABLE_NAME)
+                  .value(PK_COLUMN_NAME, STARGATE_NODE_ID)
+                  .value(VALUE_COLUMN_NAME, VALUE_COLUMN_VALUE)
+                  .ttl(INSERT_TTL_SECONDS)
+                  .timestamp(writeTimestamp.toEpochMilli())
+                  .build()
+                  .bind())
+          .get();
+
+      // select record
+      ResultSet rs =
+          dataStore
+              .execute(
+                  dataStore
+                      .queryBuilder()
+                      .select()
+                      .writeTimeColumn(VALUE_COLUMN_NAME)
+                      .from(KEYSPACE_NAME, TABLE_NAME)
+                      .where(PK_COLUMN_NAME, Predicate.EQ, STARGATE_NODE_ID)
+                      .build()
+                      .bind())
+              .get();
+
+      Row row = rs.one();
+      Instant timestampRead = Instant.ofEpochMilli(row.getLong(0));
+      if (isGreaterThanOrEqual(timestampRead, writeTimestamp)) {
+        return Result.healthy("Storage is operational");
+      } else {
+        return Result.unhealthy("Storage did not return the proper data.");
+      }
+    } catch (Exception e) {
+      logger.warn("checkIsReady failed with {}", e.getMessage(), e);
+      return Result.unhealthy("Unable to access Storage: " + e);
+    }
+  }
+
+  private boolean isGreaterThanOrEqual(Instant timestampRead, Instant expectedTimestamp) {
+    if (timestampRead == null) {
+      return false;
+    }
+    // in case there was a concurrent update, validate that the timestampRead is greater than
```

Comment:
When would this happen? There should only be one node modifying a partition at a time. Right?

Reply:
Hypothetically the health check URL might be accessed concurrently and be executed concurrently.